internal/strings/strings.go

```diff
@@ -106,3 +106,12 @@ func InSlice(a string, list []string) bool {
 func WrapUnsafe(buf []byte) string {
 	return *(*string)(unsafe.Pointer(&buf))
 }
+
+func AsciiToLower(s string) string {
+	return strings.Map(func(r rune) rune {
```

Comment:
I think this can be made even faster by working on bytes instead of runes

```go
func AsciiToLower(s string) string {
  for i := 0; i < len(s); i++ {
    c := s[i]
    if c >= 'A' && c <= 'Z' {
      return doAsciiToLower(s, i)
    }
  }
  return s
}

func doAsciiToLower(s string, start int) string {
   res := []byte(s)
   for i := start; i < len(res); i++ {
    if c >= 'A' && c <= 'Z' {
      res[i] += 32
    }
  }
  return WrapUnsafe(res)
}

Reply:
You were right, it is 3x faster

Reply:
rolled it back as it creates more allocations

Reply:
Looking at https://github.com/corazawaf/coraza/pull/790/commits/b3c7c0ab3b701c6e9a5dacedd138684962890360 you didn't follow my example which avoids examples for lowercase strings, which is necessary :P

Reply:
With your changes it generates more allocations but it was faster:
```go
# With Optimizations:
BenchmarkCRSLargePOST-10    	      13	  85454135 ns/op	 1004391 B/op	    6624 allocs/op
# Without Optimizations
BenchmarkCRSLargePOST-10    	      13	  85711750 ns/op	 1012146 B/op	    6621 allocs/op
```

```go
func AsciiToLower(s string) string {
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c >= 'A' && c <= 'Z' {
			return doAsciiToLower(s, i)
		}
	}
	return s
}

func doAsciiToLower(s string, start int) string {
	res := []byte(s)
	for i := start; i < len(res); i++ {
		c := res[i]
		if c >= 'A' && c <= 'Z' {
			res[i] += 32
		}
	}
	return WrapUnsafe(res)
}
```

Reply:
Since `BenchmarkCRSLargePost` is a e2e type of benchmark it has some varying behavior, I think we would see fairly wide variance in number of allocs across running it multiple times with the same code (I see that on my computer right now just rerunning it a few times).

It's good to look at its timing numbers mostly as a sanity check or general trend, but to confirm the allocations / true performance of `AsciiToLower` I think we can only look at the microbenchmark added in this PR. It would be good to add a case to the microbench with already lowered string too.

Reply:
My last branch contains additional benchmarks and tests, also:
```go
BenchmarkAsciiVsUnicodeCaseString/hacky_ascii-10         	 3117122	       369.9 ns/op	     128 B/op	       3 allocs/op
BenchmarkAsciiVsUnicodeCaseString/standard_ascii-10      	1000000000	         0.0000008 ns/op	       0 B/op	       0 allocs/op
BenchmarkAsciiVsUnicodeCaseString/unicode-10             	 3319794	       358.6 ns/op	     128 B/op	       3 allocs/op
```